zebra-consensus/src/transaction/check.rs

```diff
@@ -0,0 +1,165 @@
+pub fn validate_joinsplit_sig(
+    joinsplit_data: &JoinSplitData<Groth16Proof>,
+    sighash: &[u8],
+) -> Result<(), VerifyTransactionError> {
+    ed25519::VerificationKey::try_from(joinsplit_data.pub_key)
+        .and_then(|vk| vk.verify(&joinsplit_data.sig, sighash))
+        .map_err(VerifyTransactionError::Ed25519)
+}
+
+/// Calculate the Spend/Output binding signature validating key.
+///
+/// Getting the binding signature validating key from the Spend and Output
+/// description value commitments and the balancing value implicitly checks
+/// that the balancing value is consistent with the value transfered in the
+/// Spend and Output descriptions but also proves that the signer knew the
+/// randomness used for the Spend and Output value commitments, which
+/// prevents replays of Output descriptions.
+///
+/// The net value of Spend transfers minus Output transfers in a transaction
+/// is called the balancing value, measured in zatoshi as a signed integer
+/// v_balance.
+///
+/// Consistency of v_balance with the value commitments in Spend
+/// descriptions and Output descriptions is enforced by the binding
+/// signature.
+///
+/// Instead of generating a key pair at random, we generate it as a function
+/// of the value commitments in the Spend descriptions and Output
+/// descriptions of the transaction, and the balancing value.
+///
+/// https://zips.z.cash/protocol/canopy.pdf#saplingbalance
+pub fn balancing_value_balances(
+    shielded_data: &ShieldedData,
+    value_balance: Amount,
+) -> redjubjub::VerificationKeyBytes<Binding> {
+    let cv_old: ValueCommitment = shielded_data.spends().map(|spend| spend.cv).sum();
+    let cv_new: ValueCommitment = shielded_data.outputs().map(|output| output.cv).sum();
+    let cv_balance: ValueCommitment = ValueCommitment::new(jubjub::Fr::zero(), value_balance);
+
+    let key_bytes: [u8; 32] = (cv_old - cv_new - cv_balance).into();
+
+    key_bytes.into()
+}
+
+/// Check that at least one of tx_in_count, nShieldedSpend, and nJoinSplit MUST
+/// be non-zero.
+///
+/// https://zips.z.cash/protocol/canopy.pdf#txnencodingandconsensus
+pub fn some_money_is_spent(tx: &Transaction) -> Result<(), VerifyTransactionError> {
+    match tx {
+        Transaction::V4 {
+            inputs,
+            joinsplit_data,
+            shielded_data,
+            ..
+        } => {
+            if inputs.len() > 0
+                || joinsplit_data
+                    .as_ref()
+                    .map(|data| data.joinsplits().count() > 0)
+                    .unwrap_or(true)
+                || shielded_data
+                    .as_ref()
+                    .map(|data| data.spends().count() > 0)
+                    .unwrap_or(true)
+            {
+                return Ok(());
+            } else {
+                return Err(VerifyTransactionError::NoTransfer);
+            }
+        }
+        _ => return Err(VerifyTransactionError::WrongVersion(tx.to_string())),
+    }
+}
```

Comment:
Where's the corresponding rule for `tx_out_count`?

![image](https://user-images.githubusercontent.com/643204/96113851-aba14900-0edc-11eb-93cd-4befa4849451.png)

Reply:
https://github.com/ZcashFoundation/zebra/blob/transaction-verifier/zebra-consensus/src/transaction/check.rs#L68

Reply:
The function you linked only checks transparent inputs, not transparent outputs.